src/components/forms/formComponents.js

```diff
@@ -11,75 +11,74 @@ import PublishIcon from '@mui/icons-material/Publish';
 import './formComponents.css';
 import { isArray } from 'lodash';
 
+import { isValid, validateFormValue, validationMessage } from './formValidation';
+import { useState } from 'react';
+
 const styles = {
   inputStyle: {
     padding: '25px 15px',
     width: '100%'
   }
 };
 
-//---------------------------------------------
-// Form Behavior
-//---------------------------------------------
-const validateFormInput = (config, value) => {
-  const { setError, validators } = config;
-  if (validators) {
-    const validationResults = validators
-      .filter(validator => !validator.isValid(value))
-      .map(x => x.msg);
-
-    const isValid = validationResults.length === 0;
-    setError(isValid ? undefined : validationResults);
-    return isValid;
-  }
-  setError();
-  return true;
+export const getKey = ({ name, id }) => {
+  return (!isNil(name) ? name : id);
 };
 
-const getKey = (config) => {
-  return (!isNil(config.name) ? config.name : config.id);
+const updateValidation = (config, value) => {
+  const {
+    setValidation,
+    validators
+  } = config;
+  const validation = validateFormValue(value, validators);
+  setValidation(validation);
+
+  return isValid(validation);
 };
 
 const onFormInputChange = (config, value) => {
-  const { type, onChange, formValue, setFormValue } = config;
+  const { type, onChange, formValue, setFormValue, validators, setValidation } = config;
 
   const key = getKey(config);
-  const isValidInput = validateFormInput(config, value);
+
+  const validation = validateFormValue(value, validators);
+  setValidation(validation);
+
 
   if (!isNil(type?.parseFormInput)) {
     value = type.parseFormInput(value);
   }
 
   if (value !== formValue) {
-    onChange({key: key, value: value, isValid: isValidInput });
+    onChange({key: key, value: value, isValid: isValid(validation) });
     setFormValue(value);
   }
 };
 
-const errorMessage = (error) => {
-  return error && div({ className: `error-message fadein`}, [
+const errorMessages = (validation) => {
+  return !isValid(validation) && div({ className: `error-message fadein`}, [
     span({ className: 'glyphicon glyphicon-play' }),
-    ...error.map((err) => div([err])),
+    ...validation.failed.map((err) => div([validationMessage(err)])),
   ]);
 };
 
 //---------------------------------------------
 // Form Controls
 //---------------------------------------------
-export const formInputGeneric = (config) => {
+export const FormInputGeneric = (config) => {
```

Comment:
for context: these were all renamed to start with a capital letter because react only considers functions with a capital letter to be components. this matters because in one of them I am now keeping a `useState`, and React was unhappy about this.